### PR TITLE
None state is never used

### PR DESCRIPTION
**Pull request by shargon**

### CODE REVIEW

**Comment by shargon** [comment]
The problem is `StepOver`
We remove the `Break` state here: https://github.com/neo-project/neo-vm/blob/744c2dd7e92ab44bc40da4c2c7a7b190085d7707/src/neo-vm/Debugger.cs#L76
So if i remove `None` we will set the state to 0x00, and if we don't touch anything there are sites like this, impossible to be called.
https://github.com/neo-project/neo-vm/blob/744c2dd7e92ab44bc40da4c2c7a7b190085d7707/src/neo-vm/Debugger.cs#L57

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo-vm/blob/744c2dd7e92ab44bc40da4c2c7a7b190085d7707/src/neo-vm/Debugger.cs#L83-L84

When `StepOver` has done, the `BREAK` flag is added back. So the behaviour is correct. We can just do nothing with them.

File: src/neo-vm/Debugger.cs [hunk 1]
```diff
@@ -64,24 +62,23 @@ public VMState StepOut()
...
             int c = engine.InvocationStack.Count;
             do
             {
                 ExecuteAndCheckBreakPoints();
             }
             while (!engine.State.HasFlag(VMState.HALT) && !engine.State.HasFlag(VMState.FAULT) && !engine.State.HasFlag(VMState.BREAK) && engine.InvocationStack.Count > c);
-            if (engine.State == VMState.NONE)
-                engine.State = VMState.BREAK;
+
+            if (engine.State == 0x00) engine.State = VMState.BREAK;
```
Comment (erikzhang) [inline src/neo-vm/Debugger.cs]: This is not as good as giving 0x00 a name `None`.

File: src/neo-vm/Debugger.cs [hunk 2]
```diff
@@ -64,22 +62,22 @@ public VMState StepOut()
             int c = engine.InvocationStack.Count;
             while (!engine.State.HasFlag(VMState.HALT) && !engine.State.HasFlag(VMState.FAULT) && !engine.State.HasFlag(VMState.BREAK) && engine.InvocationStack.Count >= c)
                 ExecuteAndCheckBreakPoints();
-            if (engine.State == VMState.NONE)
-                engine.State = VMState.BREAK;
```
Comment (erikzhang) [inline src/neo-vm/Debugger.cs]: Noooo.... When the script is executed, it is possible to enter the `FAULT` state. We cannot directly rewrite it as `BREAK` here.

Comment (shargon) [reply src/neo-vm/Debugger.cs]: I remove both lines, so is not rewrited his state

Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: But after `StepOut()` is finished, you need to change the state to `BREAK`.

Comment (shargon) [reply src/neo-vm/Debugger.cs]: But is not possible to be `NONE`, so already is `FAULT`, `HALT` or `BREAK`

Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: L63
```c#
engine.State &= ~VMState.BREAK;
```

Comment (shargon) [reply src/neo-vm/Debugger.cs]: I was looking in the wrong method, you are right xD

File: src/neo-vm/Debugger.cs [hunk 3]
```diff
@@ -73,13 +75,15 @@ public VMState StepOver()
...
                 return engine.State;
-            engine.State &= ~VMState.BREAK;
+
+            engine.State = VMState.NONE;
             int c = engine.InvocationStack.Count;
             do
             {
                 ExecuteAndCheckBreakPoints();
             }
             while (!engine.State.HasFlag(VMState.HALT) && !engine.State.HasFlag(VMState.FAULT) && !engine.State.HasFlag(VMState.BREAK) && engine.InvocationStack.Count > c);
```
Comment (shargon) [inline src/neo-vm/Debugger.cs]: Check only none?

File: src/neo-vm/Debugger.cs (hunk 2 above)
Comment (shargon) [reply src/neo-vm/Debugger.cs]: But check this
https://github.com/neo-project/neo-vm/pull/185/files#diff-ee43c09353d045e18dc0046eb81d86dfR66

Comment (shargon) [reply src/neo-vm/Debugger.cs]: One of my problems is that i can touch this line with ut https://codecov.io/gh/neo-project/neo-vm/src/master/src/neo-vm/Debugger.cs

Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: Don't understand.

File: src/neo-vm/Debugger.cs (hunk 3 above)
Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: No. Each step has the potential to change the state of the VM.

File: src/neo-vm/Debugger.cs [hunk 4]
```diff
@@ -33,7 +33,7 @@ public VMState Execute()
         private void ExecuteAndCheckBreakPoints()
         {
             engine.ExecuteNext();
-            if (engine.State == VMState.NONE && engine.InvocationStack.Count > 0 && break_points.Count > 0)
+            if (engine.InvocationStack.Count > 0 && break_points.Count > 0)
```
Comment (erikzhang) [inline src/neo-vm/Debugger.cs]: You can't remove the condition `engine.State == VMState.NONE`.

File: src/neo-vm/Debugger.cs [hunk 5]
```diff
@@ -53,17 +53,19 @@ public VMState StepInto()
             if (engine.State.HasFlag(VMState.HALT) || engine.State.HasFlag(VMState.FAULT))
                 return engine.State;
             engine.ExecuteNext();
-            if (engine.State == VMState.NONE)
-                engine.State = VMState.BREAK;
```
Comment (erikzhang) [inline src/neo-vm/Debugger.cs]: You can't remove these two lines.

The last line:

```c#
engine.ExecuteNext();
```

It is possible to cause the VM to enter the FAULT state.

File: src/neo-vm/Debugger.cs [hunk 6]
```diff
@@ -53,17 +53,19 @@ public VMState StepInto()
...
             return engine.State;
         }
 
         public VMState StepOut()
         {
-            engine.State &= ~VMState.BREAK;
+            if (engine.State.HasFlag(VMState.HALT) || engine.State.HasFlag(VMState.FAULT))
+                return engine.State;
+            engine.State = VMState.NONE;
```
Comment (erikzhang) [inline src/neo-vm/Debugger.cs]: No difference.

File: src/neo-vm/Debugger.cs [hunk 7]
```diff
@@ -73,13 +75,15 @@ public VMState StepOver()
         {
             if (engine.State.HasFlag(VMState.HALT) || engine.State.HasFlag(VMState.FAULT))
                 return engine.State;
-            engine.State &= ~VMState.BREAK;
+
+            engine.State = VMState.NONE;
```
Comment (erikzhang) [inline src/neo-vm/Debugger.cs]: No difference.

File: src/neo-vm/Debugger.cs (hunk 5 above)
Comment (shargon) [reply src/neo-vm/Debugger.cs]: but fault is not none, i can't touch this line with unit test...

File: src/neo-vm/Debugger.cs (hunk 3 above)
Comment (shargon) [reply src/neo-vm/Debugger.cs]: but this is the same as equal none

Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: So?

File: src/neo-vm/Debugger.cs (hunk 5 above)
Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: I don't understand what do you mean?

File: src/neo-vm/Debugger.cs [hunk 8]
```diff
@@ -73,13 +75,15 @@ public VMState StepOver()
...
             engine.State &= ~VMState.BREAK;
             int c = engine.InvocationStack.Count;
             do
             {
                 ExecuteAndCheckBreakPoints();
             }
             while (!engine.State.HasFlag(VMState.HALT) && !engine.State.HasFlag(VMState.FAULT) && !engine.State.HasFlag(VMState.BREAK) && engine.InvocationStack.Count > c);
+
             if (engine.State == VMState.NONE)
                 engine.State = VMState.BREAK;
```
Comment (shargon) [inline src/neo-vm/Debugger.cs]: @erikzhang how can i enter in this condition?

File: src/neo-vm/Debugger.cs (hunk 5 above)
Comment (shargon) [reply src/neo-vm/Debugger.cs]: @erikzhang how can i enter in this condition?

File: src/neo-vm/Debugger.cs (hunk 3 above)
Comment (shargon) [reply src/neo-vm/Debugger.cs]: 3 checks vs one, is better, state is not a flag now

Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: `engine.State != VMState.None`?

File: src/neo-vm/Debugger.cs (hunk 5 above)
Comment (erikzhang) [reply src/neo-vm/Debugger.cs]: Call `StepInto()` when vm is in any non-fault state.

File: src/neo-vm/Debugger.cs (hunk 3 above)
Comment (shargon) [reply src/neo-vm/Debugger.cs]: yes

**Comment by shargon** [comment]
I will remove the flag checks and back here again

**Pull request closed without merging**